pkg/controller/volume/attach_detach_controller.go

```diff
@@ -19,63 +19,109 @@ limitations under the License.
 
-type attachDetachController struct {
-	// internalPodInformer is the shared pod informer used to fetch and store
-	// pod objects from the API server. It is shared with other controllers and
-	// therefore the pod objects in its store should be treated as immutable.
-	internalPodInformer framework.SharedInformer
+	// SafeToDetachAnnotation is the annotation added to the Node object by
+	// kubelet in the format "volumes.kubernetes.io/safetodetach/{volumename}"
+	// to indicate the volume has been unmounted and is safe to detach.
+	SafeToDetachAnnotation string = "volumes.kubernetes.io/safetodetach/"
 
-	// selfCreatedPodInformer is true if the internalPodInformer was created
-	// during initialization, not passed in.
-	selfCreatedPodInformer bool
+	// loopPeriod is the ammount of time the reconciler loop waits  between
+	// successive executions
+	reconcilerLoopPeriod time.Duration = 100 * time.Millisecond
 
-	// internalNodeInformer is the shared node informer used to fetch and store
-	// node objects from the API server. It is shared with other controllers
-	// and therefore the node objects in its store should be treated as
-	// immutable.
-	internalNodeInformer framework.SharedInformer
+	// reconcilerMaxSafeToDetachDuration is the maximum ammount of time the
+	// attach detach controller will wait for a volume to be safely detached
+	// from its node. Once this time has expired, the controller will assume the
+	// node or kubelet are unresponsive and will detach the volume anyway.
+	reconcilerMaxSafeToDetachDuration time.Duration = 10 * time.Minute
+)
 
-	// selfCreatedNodeInformer is true if the internalNodeInformer was created
-	// during initialization, not passed in.
-	selfCreatedNodeInformer bool
+// AttachDetachController defines the operations supported by this controller.
+type AttachDetachController interface {
+	Run(stopCh <-chan struct{})
 }
 
 // NewAttachDetachController returns a new instance of AttachDetachController.
 func NewAttachDetachController(
 	kubeClient internalclientset.Interface,
 	podInformer framework.SharedInformer,
 	nodeInformer framework.SharedInformer,
-	resyncPeriod time.Duration) AttachDetachController {
+	pvcInformer framework.SharedInformer,
+	pvInformer framework.SharedInformer,
+	resyncPeriod time.Duration,
+	cloud cloudprovider.Interface,
+	plugins []volume.VolumePlugin) (AttachDetachController, error) {
 	selfCreatedPodInformer := false
 	selfCreatedNodeInformer := false
+	selfCreatedPVCInformer := false
+	selfCreatedPVInformer := false
+	// TODO: The default resyncPeriod for shared informers is 12 hours, this is
+	// unacceptable for the attach/detach controller. For example, if a pod is
+	// skipped because the node it is scheduled to didn't set its annotation in
+	// time, we don't want to have to wait 12hrs before processing the pod
+	// again.
+	// Luckily https://github.com/kubernetes/kubernetes/issues/23394 is being
```

Comment:
So looks like the only thing I have to do is update the [MinResyncPeriod](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/options/options.go#L66). However since I am using shared informers this change will affect other controllers that use the shared informers. Therefore, I will do it in a separate/isolated/explicit PR saying "I'm changing the default Resync period for shared informers from from 12 hours to 1 minute, because PR #24142 made it so Resync does not result in relist".

For this PR I will leave this comment.